### ISSUE TITLE
damaged file contents when using the loopback example

### ISSUE THREAD

**Issue by N-Coder** [billziss-gh/fusepy#8]
First off, thanks for your great work on making FUSE available on windows - in deed with minimal changes to existing code relying on fusepy!

But unfortunately, I couldn't get the loopback.py example to run without issues on windows.
The following functions of the `os` module, that are used in the official loopback example, are not available on windows: `chown, fdatasync, mknod, statvfs`
Commenting out their usage in the example (see [here](https://github.com/N-Coder/fusepy/commit/6a53bd944e043a38384424b1d8bfe5129a652527)) will make it run without an ImportError, but a comment on how to implement the now missing functionality - or even an updated loopback example using the correct functions when running on windows - would be highly appreciated!
Additionally, one should mention that all path names sent to the driver will be upper-case, probably due to Window's case-insesitivity regarding file names.

With the "fixed" loopback example, reading a bigger file now returns some of its contents damaged (as far as I can see they are replaced by nulls).
I saved some reference data as test.txt in the root folder and tried viewing or copying the file through the mounted loopback folder.
You can find the data in the attached [loopback.zip](https://github.com/billziss-gh/fusepy/files/1778374/loopback.zip) as reference-data.txt, containing the numbers 0 to 1521558 in ASCII code, each on a separate line (so that they are be lined up with their respective line number).

In my first test, I opened the file in the loopback mount using notepad++ and copy-pasted the data to a file outside of the mount. The resulting data and the output of the python code can be found in test1-data.txt and test1-log.txt, respectively. When you scroll to the very end, you can cleary see that the lines are no longer lined up with their line-numbers and some lines are empty or contain only part of the number.

In the second test, I directly copyied the file from inside the mount to an outside folder using Windows Explorer. Now, the line numbers in test2-data.txt seem to match, but there are a lot of null-bytes preceding the number in the last line.

I had similar problems when developing the fusepy driver on Linux, usually they were related to race conditions between the seek and read. As the lookback example protects it's read method using a Lock and the problem persists when calling the FUSE driver with `nothreads=True`, this is probably not a fault of the python script in this case.

I also found the following possibly related lines occuring multiple times in sequence in the log files:
```
DEBUG:fuse.log-mixin:-> open C:\Users\niko\Documents\root/TEST.TXT (0,)
DEBUG:fuse.log-mixin:<- open 4
DEBUG:fuse.log-mixin:-> flush C:\Users\niko\Documents\root/TEST.TXT (4,)
DEBUG:fuse.log-mixin:-> getattr C:\Users\niko\Documents\root/TEST.TXT (None,)
DEBUG:fuse.log-mixin:<- flush '[Errno 9] Bad file descriptor'
DEBUG:fuse.log-mixin:-> release C:\Users\niko\Documents\root/TEST.TXT (4,)
DEBUG:fuse.log-mixin:<- getattr {'st_atime': 1520157090.806919, 'st_ctime': 1520157090.806919, 'st_gid': 0, 'st_mode': 33206, 'st_mtime': 1520157117.6736915, 'st_nlink': 1, 'st_size': 12582921, 'st_uid': 0}
DEBUG:fuse.log-mixin:<- release None
```

So, this issue actually contains 3 smaller and possibly related issues:
- [ ] the loopback example currently doesn't work on windows, as it uses POSIX-only functions from `os`. A fixed loopback.py example with replacements and some comments with details on the WinFsp FUSE API (e.g. how the replacements work, and that all paths are upper-case) would be nice.
- [ ] when reading file data through the loopback mount, some of it gets corrupted and replaced by null-bytes.
- [ ] those strange open-read-release cycles, that happen multiple times in sequence

**Comment by N-Coder** [comment]
Oh, and here are the details on my test environment:
- Windows 10 Pro, Version 1709, Build 16299.248
- WinFsp 1.2.17346
- Python 3.6.4
- your fusepy fork
- the newer loopback.py example from the official repository (with my fixes), called from PyCharm IDE
- root directory/file: C:\Users\niko\Documents\root\text.txt
- mount directory/file: C:\Users\niko\Documents\mount\text.txt

**Comment by billziss-gh** [comment]
@N-Coder I apologize for my late response. I am currently busy with other work.

> The following functions of the os module, that are used in the official loopback example, are not available on windows: chown, fdatasync, mknod, statvfs
Commenting out their usage in the example (see here) will make it run without an ImportError, but a comment on how to implement the now missing functionality - or even an updated loopback example using the correct functions when running on windows - would be highly appreciated!

As you know Python does not provide these functions in their OS modules. `Mknod` makes little sense on Windows (at least for special files); `chown`, `fdatasync` and `statvfs` are doable, using `ctypes` or straight C.

(Aside: I have contemplated for some time now to write a small POSIX library that would facilitate some of these calls on Windows.)

The good news is that the file system should still be rather functional without these calls. `Statvfs` would be nice to report accurate file system information, and may be relatively easy: [`GetDiskFreeSpace`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364935(v=vs.85).aspx), [`GetDiskFreeSpaceEx`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364937(v=vs.85).aspx).

> With the "fixed" loopback example, reading a bigger file now returns some of its contents damaged (as far as I can see they are replaced by nulls).

I am not aware of any such issue on the WinFsp-FUSE layer, which is tested rather well (this particular issue would be caught by a testing tool called [fsx](https://github.com/billziss-gh/secfs.test/tree/master/fstools/src/fsx)). So the issue may be with the fusepy fork or the loopback example.

This definitely needs investigating, although I doubt that I would have the time to do it in the near future.

> I also found the following possibly related lines occuring multiple times in sequence in the log files:
> ```
> DEBUG:fuse.log-mixin:-> open C:\Users\niko\Documents\root/TEST.TXT (0,)
> DEBUG:fuse.log-mixin:<- open 4
> DEBUG:fuse.log-mixin:-> flush C:\Users\niko\Documents\root/TEST.TXT (4,)
> DEBUG:fuse.log-mixin:-> getattr C:\Users\niko\Documents\root/TEST.TXT (None,)
> DEBUG:fuse.log-mixin:<- flush '[Errno 9] Bad file descriptor'
> DEBUG:fuse.log-mixin:-> release C:\Users\niko\Documents\root/TEST.TXT (4,)
> DEBUG:fuse.log-mixin:<- getattr {'st_atime': 1520157090.806919, 'st_ctime': 1520157090.806919, 'st_gid': 0, 'st_mode': 33206, 'st_mtime': 1520157117.6736915, 'st_nlink': 1, 'st_size': 12582921, 'st_uid': 0}
> DEBUG:fuse.log-mixin:<- release None
> ```

This is actually normal on Windows.

In Windows every file operation needs to happen on an open file. These includes operations like `stat`, `unlink`, `rename`, etc. which in POSIX work on file paths rather than file descriptors.

If an application issues the Windows equivalent of a `stat` on a file in your file system, what you will see is an `open`, `getattr`/`fgetattr` and `release` on your file system. Same with `unlink`, `rename`, etc!

For this reason I recommend to avoid doing any heavy processing in `open` in a FUSE file system. Chances are Windows/WinFsp might just issue a `getattr`/`fgetattr` and then close your file. If you fetch metadata in `open` cache them, and then return them in `fgetattr` (which should really be implemented on Windows).

**Comment by N-Coder** [comment]
> I am not aware of any such issue on the WinFsp-FUSE layer, which is tested rather well (this particular issue would be caught by a testing tool called fsx). So the issue may be with the fusepy fork or the loopback example.

The read method of the "official" [loopback.py example](https://github.com/billziss-gh/fusepy/blob/master/examples/loopback.py) is actually very straightforward, as it just loops the read back to a real file on another filesystem:
```python
    def read(self, path, size, offset, fh):
        with self.rwlock:
            os.lseek(fh, offset, 0)
            return os.read(fh, size)
```
So it's either some strange peculiarity about `lseek` and `read` on Windows that leads to the seemingly timing-based data corruption or some problem of this fusepy fork when interfaceing with the specific WinFsp FUSE API, as the example works flawlessly on Linux.

> If an application issues the Windows equivalent of a stat on a file in your file system, what you will see is an open, getattr/fgetattr and release on your file system. Same with unlink, rename, etc!

The thing that seemed weird about the line in the log was the `flush '[Errno 9] Bad file descriptor'` right after opening the fd, which should actually be valid at that point (having been opened right before). I thought this could be a hint at a timing problem possibly related to the above.

**Comment by billziss-gh** [comment]
I just realize that I have an additional fix (eab1b2aca692e84c73fa9e418fd058b6bf8a15d3) in the Windows branch. Please give that a shot and see if it fixes your issue.

**Comment by N-Coder** [comment]
I tried it with the different branch, but no luck there, either.

After some more fiddling around, I found out that the problem seemed to be due to the `os.read` call not returning enough data (even when calling it multiple times in a loop, it would start to return 0 bytes half way through the requested range). In the end, the problem seemed to be how the loopback example opens the file, by directly reusing the `os.open` function: [`open = os.open`](https://github.com/billziss-gh/fusepy/blob/master/examples/loopback.py#L56).
By default, the `flags` parameter seems to be 0, or `O_RDONLY` without `O_BINARY` or `O_TEXT` set. If you manually set it to include `O_BINARY`, everything works fine:
```python
    def open(self, path, flags):
        return os.open(path, os.O_RDONLY | os.O_BINARY)
```
